text/0000-generic-dependent-consts.md

```diff
@@ -0,0 +1,275 @@
+constant `C` is defined, `EXPR` is required to be a constant expression, and can
+be used in certain situations (especially patterns) where inlining the
+expression, either manually or via macro, would produce invalid syntax. (Of
+course, named constants also improve readability and allow for clearer error
+messages in some cases.)
+
+Keeping this in mind, it is straightforward to understand most uses of constant
+expressions. The design below will mostly focus on special cases and
+interactions with other language features.
+
+## Other items in functions
+
+The scope of this RFC is limited to consts. Nested functions will still be
+forbidden from referencing "outer" type parameters. The same is true for
+statics. There are two justifications for this.
+
+Firstly, nested static items, whether they are static variables or functions,
+must have static locations in memory. If these items could use the type
+parameters from the surrounding scope, they would have to be newly instantiated
+for each instantiation of the enclosing function.
+
+Secondly, functions that make use of outer type parameters are in effect higher-
+kinded types. These seem to require more careful consideration to implement.
+
+Neither of these considerations apply to constants, which the compiler is always
+allowed (and often required) to inline. They do not need a static location in
+memory, and unlike functions, they cannot add new type or lifetime parameters to
+those already present in the surrounding scope.
+
+## `'static` references
+
+Consider the following definitions of similar functions:
+
+```rust
+// Currently not valid (borrowed value does not live long enough).
+fn ref_one_literal() -> &'static u32 {
+    &1
+}
+// Currently valid (a static is implicitly created).
+fn ref_one_const() -> &'static u32 {
+    const REF_ONE: &'static u32 = &1;
+    REF_ONE
+}
+// Not valid (borrowed value does not live long enough).
+fn ref_one_ref_const() -> &'static u32 {
+    const ONE: u32 = 1;
+    &ONE
+}
+// Should either of the following become valid?
+fn ref_one_generic_const<T: Int>() -> &'static T {
+    const REF_ONE: &'static u32 = &T::ONE;
+    REF_ONE
+}
+fn ref_one_associated_const<T: Int>() -> &'static T {
+    &T::ONE
+}
+```
+
+This RFC proposes that both `ref_one_generic_const` and
+`ref_one_associated_const` would be invalid.
+
+The associated const case is disallowed for the same reason as
+`ref_one_ref_const`, i.e. because the expression is not `'static` and thus does
+not live long enough for a static borrow.
+
+The generic const case is disallowed because it implicitly creates a static item
+that depends on a type parameter, and this is forbidden, as mentioned in the
+preceding section. That is to say, the initializer expression itself is invalid.
+
+In order to implement this restriction, use of a type parameter in a constant
+expression must be considered "contagious". That is, if the initializer
```

Comment:
I don't think it matters anymore because it's been explained to me. Only if the misunderstanding is common should be it changed. I don't have any evidence that it is.